Top header animation

VortexExpansion opened the pull request:
@HisashiHasebe @yabe-diverta san

With reference to https://github.com/diverta/front_next_media/pull/3#issuecomment-1884366167
I created a new PR because this needed a change in architecture of folders and has affected a lot of files for a small change - implementing the animation in header. 

I hence request you to review this separately from my main PR. 

@KuriharaMina san, I request you to please take a look at the implementation of animation in headers in this PR since it is also a design oriented task. Please let me know if we can implement an easier solution for the same.

Link : https://diverta.gyazo.com/c376c799148c8663d09f4b1d8b287624

VortexExpansion committed:
fix warnings

VortexExpansion committed:
add cardview

VortexExpansion committed:
list view for all pages

VortexExpansion committed:
add details page

VortexExpansion committed:
Merge remote-tracking branch 'upstream/main' into next-media

VortexExpansion committed:
update details page

VortexExpansion committed:
upd pagetitle

VortexExpansion committed:
fix layout-template render for categories

VortexExpansion committed:
add feature page

VortexExpansion committed:
upd feature detail body

VortexExpansion committed:
upd privacy page

VortexExpansion committed:
clean code

VortexExpansion committed:
export barrel

VortexExpansion committed:
upd privacy policy text

VortexExpansion committed:
implement pager

VortexExpansion committed:
upd getContentList path and layout

VortexExpansion committed:
migrate pagination to article only

VortexExpansion committed:
filter by tags

VortexExpansion committed:
implement keyword search

VortexExpansion committed:
unify searchQueryParams behaviour into article page

VortexExpansion committed:
refactor categories into ArticlePage as query

VortexExpansion committed:
upd pager and breadcrumb

VortexExpansion committed:
rmv env from commit

VortexExpansion committed:
fix output:export

VortexExpansion committed:
add ranking; upd fav_cnt

VortexExpansion committed:
search bugfix

VortexExpansion committed:
fix pagination

VortexExpansion committed:
minor link fixes

VortexExpansion committed:
fix detail page view

VortexExpansion committed:
Top Header animation

VortexExpansion commented:
@yabe-diverta  san 

I request you to review and merge https://github.com/diverta/front_next_media/pull/3 if you have time. 

Since none of my PRs are merged yet, it is making it difficult for me to create PRs for independent isolated tasks that might have potential problems or are tough to implement because all commits are carried forward. It would make it difficult to track progress and review independent functionalities as well. 🙇

HisashiHasebe commented:
@VortexExpansion 
- Can you check this bug?
I couldn't back to home from detail page.
https://diverta.gyazo.com/04ddc77cacf1b56cef82f67cbbb6e428

- And please update detail page
Please show "記事詳細" instead of category name.
https://diverta.gyazo.com/0975b4be2590058c030d71259951d118

VortexExpansion committed:
upd detailBody

VortexExpansion committed:
downgrade next to 13.4.10

VortexExpansion commented:
@HisashiHasebe san

Details path bug fixed https://diverta.gyazo.com/335af33771bd31653a4744d548ccf0f5

The returning to home bug with error in console 
`error TypeError [ERR_INVALID_STATE]: Invalid state: ReadableStream is already closed`
again seems to be a bug in the newer versions of next (`>13.4.10`). 
Downgrading next to `13.4.10` seems to solve the problem

Reference : https://github.com/vercel/next.js/discussions/55027#discussioncomment-6947969

P.S We need to careful about the latest versions of Next 13. I had to downgrade a couple of times now due to potential version bugs. @yabe-diverta

HisashiHasebe commented:
@VortexExpansion 
Thank you for fix.
Looks and functions is good for me.

@yabe-diverta 
Please proceed code review.

KuriharaMina commented:
@VortexExpansion 
I believe the movement of the headers on the top and subpages is fine. 
Thank you for the corrections.

HisashiHasebe commented:
@yabe-diverta 
Check code please

yabe-diverta commented on .env:
```diff
@@ -0,0 +1 @@
+NEXT_PUBLIC_BASE_URL=https://dev-next-media.g.kuroco.app
```
@VortexExpansion 
Next js does not recommend to use simple `.env` file, pls rename it.

@HisashiHasebe @VortexExpansion 
I think we once before discussed how we should define env variables at building Nuxt3 time,
But I'm now on memory loss 😢 
Is renaming `.env` to `.env.development` desirable?

HisashiHasebe replied:
@yabe-diverta @VortexExpansion 
nuxt.config.ts に書いています。envにする必要があるなら `.env.example` ですね。
https://github.com/diverta/front_nuxt_corporate/pull/1#issuecomment-1776384675

yabe-diverta replied:
ありがとうございます。
runtimeConfigはNextでは非推奨にされているので、`.env.example`ですね。

@VortexExpansion 
pls rename `.env` to `.env.example`, then copy it as `.env.local` 🙇

yabe-diverta commented on .gitignore:
```diff
@@ -26,6 +26,7 @@ yarn-error.log*
 
 # local env files
 .env*.local
+.env
```
@VortexExpansion 
No need, pls remove :bow:

yabe-diverta commented on package.json:
```diff
@@ -9,7 +9,7 @@
   },
   "dependencies": {
-    "next": "latest",
+    "next": "13.4.10",
```
@VortexExpansion 
v13.4 is a bit old, please consider upgrading to the latest ([v14.0.4](https://github.com/vercel/next.js/releases/tag/v14.0.4)?)

VortexExpansion replied:
@yabe-diverta san

Done 🎉

yabe-diverta commented on src/components/layouts/Header.jsx:
```diff
@@ -1,10 +1,41 @@
+  const dataJsAttribute = topPage.topPage ? 'header-scroll' : null;
+
+  useEffect(() => {
+    const header = document.querySelector('[data-js="header-scroll"]');
```
@VortexExpansion 
Do not use `js-...` special class, use `useRef()` instead.

yabe-diverta commented on src/components/common/index.js:
```diff
@@ -0,0 +1,5 @@
+export { default as Breadcrumb } from './Breadcrumb';
+export { default as PageTitle } from './PageTitle';
+export { default as TagArea } from './TagArea';
+export { default as TagKeyword } from './TagKeyword';
```
@VortexExpansion 
`index.js` is a bit old solution because TreeShaking probably does not work properly.
I think just using `export function Banner()...` in each component is appropriate.

yabe-diverta commented on src/components/common/fetchData.js:
```diff
@@ -0,0 +1,88 @@
+    var url = `${process.env.NEXT_PUBLIC_BASE_URL}/rcms-api/1/content/list?pageID=${pageID}&contents_type=${categoryID}&tag_id=${tag_id}`;
+    if(search) url+=`&filter=keyword%20contains%20${search}`;
+    console.log(url);
+    const res = await fetch(url , { cache: "no-store" })
```
@VortexExpansion 
In our usecase we just simple output static HTMLs in most cases, please consider setting cache instead.
Because we just once fetch data from API at SSG(server) process, no need to get fresh data in similar pages/components.
(For dynamic contents we should avoid using cache in opposite)
https://nextjs.org/docs/app/building-your-application/data-fetching/fetching-caching-and-revalidating

VortexExpansion replied:
@yabe-diverta san

I think just removing `{ cache: "no-store" }` should be sufficient to cache all data then. 

I don't think there is any dynamic content apart from adding to favourite lists at the moment. I will try to figure out a way to handle updating those dynamically as a special case. 
(the dynamic content for card view article/[id] has different paths so should not be a problem)

Is this OK ?

yabe-diverta replied:
Sure, thank you to check it.

yabe-diverta commented on src/app/(pages)/(TwoColumn)/article/[id]/page.jsx:
```diff
@@ -0,0 +1,39 @@
+export default async function Food({ params }) {
+  const item = await getDetails(params.id);
+
+  const content = {
```
@VortexExpansion 
please use `useMemo()`

VortexExpansion replied:
@yabe-diverta san

According to https://nextjs.org/docs/app/building-your-application/caching#request-memoization

Do you think we can solve the problem of memoizing by just removing the `{ cache: "no-store" }` line from every fetch command inside [fetchData.js](https://github.com/diverta/front_next_media/pull/5/files#diff-5b3a7c3287ce696355fe5bbb747e3259da918166422a94978e2e032d8f655130R21) and not using the react useMemo hook since memoizing the fetch calls is a default behaviour in Next 13 

What do you think ? Am I missing something ? 🙇

yabe-diverta replied:
No it's not fetch stuff.
I meant it was not a good way to define object in a component,
because it would consume memory on every re-rendering.

like:
```js
  const content = {
    text: '記事詳細',
    text_en: 'Article details',
  };
```
should be:
```js
  const content = useMemo({
    text: '記事詳細',
    text_en: 'Article details',
  }, []);
```
useMemo() or useRef() to wrapup data.

The `content` data seems too light weight to consider memory usage in a sense,
although wrapping up is standard way for many React ppl.
So it's easier to wrap such any data in every components rather than thinking deeply either is better in every components.

VortexExpansion replied:
@yabe-diverta san

Same as https://github.com/diverta/front_next_media/pull/5#discussion_r1456779076
What I tried : `src/app/(pages)/(TwoColumn)/article/[id]/page.jsx` replaced by code in [page.txt](https://github.com/diverta/front_next_media/files/13971821/page.txt)

Using useMemo would require changing to client component from server. 
Entering infinite loop of console statements due to multiple re-renders. 

Would prefer leaving it as a server component because I believe the content is not a big overhead and is not rendered multiple times during navigation

Once I load any of the pages let's say article/1 it is stored in the cache by default, if I go to article/2 and then come back to article/1 it will not fetch the content value or the api again but rather just load it from the cache. So using useMemo and converting it to client should not be necessary according to me. 

Please check and let me know what you think. 🙇

yabe-diverta replied:
> Using useMemo would require changing to client component from server.

Yes it's not good to change to client component as well.
please ignore my comment 🙇

yabe-diverta commented on src/app/(pages)/(TwoColumn)/article/page.jsx:
```diff
@@ -0,0 +1,68 @@
+} from "@/components/common";
+
+export default async function Event({ searchParams }) {
+  const page = Number(
```
@VortexExpansion 
pls use `useCallback()`

VortexExpansion replied:
@yabe-diverta san 

I have a concern here. To use useCallback I would have to force this page to be a client component whereas currently this is a server component. Would that not also be less efficient than before.

Secondly, I had some problems using useCallback here because the default value I provided was being re-rendered again and again in the useEffect leading to infinite console statements. 

I am attacking a txt file for `src/app/(pages)/(TwoColumn)/article/page.jsx` 
can you please replace this and check how to proceed ahead.

[page.txt](https://github.com/diverta/front_next_media/files/13971693/page.txt)

yabe-diverta replied:
> currently this is a server component

Understood,,, sry this suggestion was due to my ignorant for Next AppRouter 🙇 
please ignore my comment.

yabe-diverta commented on src/app/(top)/page.jsx:
```diff
@@ -1,114 +1,67 @@
-                  <p className="c-card__category">CULTURE</p>
+          {topRankedList.map((item, index) => (
+            <li className="c-card__item" key={index}>
+              <a href={`/article/${item.topics_id}`} className="c-card">
```
@VortexExpansion 
pls use `<NextLink>` for internal links

yabe-diverta commented on src/components/common/Pager.jsx:
```diff
@@ -1,16 +1,64 @@
+        <>
+          <li className="c-pager__item">{navigateToPage(1, "&laquo;")}</li>
+          <li className="c-pager__item">
+            {navigateToPage(page - 1, "&lsaquo;")}
```
@VortexExpansion 
`"&lsaquo;"` seems somewhat magic number,
pls consider below either:
- define `constants.js` and name them better
- define wrapper function with better naming

yabe-diverta commented on src/components/common/fetchData.js:
```diff
@@ -0,0 +1,88 @@
+export async function getContentList(contentCategory, pageID = 1, tag_id="", search="") {
```
@VortexExpansion 
`tag_id="",` seems not processed with Prettier, pls check your formatter config?

VortexExpansion replied:
@yabe-diverta  san

Does not show up on my prettier formatter. But changed default value to `tag_id = null` instead. 

Can you confirm now ?

yabe-diverta replied:
Hmm reconsidered, it would be better to set up the config in PJ 🤔 
how do you think?

VortexExpansion replied:
@yabe-diverta san

I added a template .`prettierrc` file https://github.com/diverta/front_next_media/pull/5/commits/0e4f78b341f4fd8ab37053d1410b8390835c768b

Please feel free to add any of the content you think is important and missing. 

What do you think about this ?

yabe-diverta replied:
Yes it's sufficient.
closing this conversation.

yabe-diverta commented on src/components/section/feature/FeatureDetailBody.jsx:
```diff
@@ -0,0 +1,46 @@
+        {data.ext_2}
+      </div>
+      <ul className="c-feature__list">
+        {data.ext_3.map(async (item, index) => {
```
@VortexExpansion 
Pls avoid using async fn as an inline one due to causing heavy performance issue.
Use Promise.all() and `useMemo()/useCallback()` instead.

VortexExpansion replied:
@yabe-diverta san

I used Promise.all and useCallBack and removed it from being inline 
https://github.com/diverta/front_next_media/pull/5/commits/2308cdeaa1ac6cb3a865bda7e6716af071ebe648

But I noticed that the performance is now worse and it takes more time to fetch and load. Can you please review this once 
https://diverta.gyazo.com/b40e8ea6ebd41613ad5c3d44ebeceb60

I felt that the load times of the code before was much faster https://diverta.gyazo.com/6ab6ba2db39b9688b16d0072792f9e0c

yabe-diverta replied:
Far more checking, I think this client component should not fetch data, changed to fetch data at server and pass the data as props.
pls check [this update](https://github.com/diverta/front_next_media/pull/5/commits/2238cea1d4d0deda329925eca6c71a344da1b4c6) 🙇

yabe-diverta commented on tsconfig.json:
```diff
@@ -1,7 +1,11 @@
 {
```
@VortexExpansion 
Can we remove this?

VortexExpansion replied:
@yabe-diverta san

Refactored to jsconfig.json https://github.com/diverta/front_next_media/pull/5/commits/5d963366a47a493354307854d7fe85e10cd4a766

Please review. 🙇

yabe-diverta replied:
looks good, thank you 🙇

yabe-diverta commented on src/components/ui/Search.jsx:
```diff
@@ -0,0 +1,33 @@
+            placeholder="キーワードを入力"
+            className="l-header__nav__searchText"
+            value={searchTerm}
+            onChange={handleInputChange}
```
@VortexExpansion 
Pls consider debouncing

VortexExpansion replied:
@yabe-diverta san

The search here is not dynamic on every key stroke but are fetched only when the user presses the submit button.
Only the searchTerm is updated on every keystroke but the value is fetched only once after pressing submit so I think debounce here would not be a significant performance boost. 

Do you think implementing debounce here is still important ? 🙇

yabe-diverta replied:
Ah understood.
No need to implement it, thank you for clarification 🙇

yabe-diverta reviewed (changes requested):
@VortexExpansion 
Reviewed, sry so many points I raised 🙇 
Pls check.

VortexExpansion committed:
add constants.js

VortexExpansion committed:
update next links

VortexExpansion committed:
update imports; rmv index.js

VortexExpansion committed:
update env

VortexExpansion committed:
fix prettier

VortexExpansion committed:
login bugfix for build

VortexExpansion committed:
rmv data-js

VortexExpansion committed:
remove no-store cache for fetch requests

VortexExpansion committed:
replace inline async with promise + useCallback in featureDetails

VortexExpansion committed:
add .prettierrc

VortexExpansion committed:
rmv ts files

VortexExpansion committed:
upgrade to next 14.0.4

VortexExpansion commented:
@HisashiHasebe san I would like to request a review from you again for testing all the pages once again because I have updated the Next version to `14.0.4` based on review by @yabe-diverta san 

If something breaks, please let me know. As far as I debugged, everything seems to work fine in this version after adding [generateStaticParams](https://nextjs.org/docs/app/api-reference/functions/generate-static-params).

P.S 
`member/mypage` won't work because auth is implemented in PR#4.
`npm run build` still won't work for searchParams but works for all other pages https://diverta.gyazo.com/da4b5949590840fdd6156e5e6c3318bd

@yabe-diverta san 
I have addressed all your reviews. Please review again 🙇

HisashiHasebe commented:
@VortexExpansion 
An error occurs.
Can you fix it?
https://diverta.gyazo.com/1b08b4d5443766b481f9c79630f5965c

VortexExpansion commented:
@HisashiHasebe san 

Sorry, I think the reason might be that I have reset the environment. 

Please create a file `.env.local` and set value `NEXT_PUBLIC_BASE_URL=https://dev-next-media.g.kuroco.app`
like https://diverta.gyazo.com/1d209dcc729ccad361153d987289cc02

HisashiHasebe commented:
@VortexExpansion 
Oh, Sorry work well now.
And looks good for me!

VortexExpansion committed:
Merge remote-tracking branch 'upstream/main' into top-header

VortexExpansion committed:
fix conflicts

yabe-diverta committed:
chore: consider empty data in CardList

yabe-diverta committed:
Merge branch 'top-header' of https://github.com/VortexExpansion/front_next_media into top-header

yabe-diverta committed:
refactor: fetching data at server

yabe-diverta committed:
fix: invalid fetch param

yabe-diverta committed:
feat: change param for Breadcrumbs

yabe-diverta committed:
feat: edit ext_n field props

yabe-diverta committed:
feat: set alt text

yabe-diverta commented:
@VortexExpansion 
Thank you to react all the comments 🙇 , now all resolved, please merge this PR if it's OK.

HisashiHasebe commented:
@VortexExpansion 
Sorry, Is it front matter?
https://diverta.gyazo.com/3cb96ad4d676f2860bf3f67b4d66eb6f

yabe-diverta commented:
@HisashiHasebe 

> @VortexExpansion Sorry, Is it front matter? https://diverta.gyazo.com/3cb96ad4d676f2860bf3f67b4d66eb6f

It's not frontend issue, I set all the ext_n columns to special slug.
I'm not sure how did you reproduce it, but the error should not show while API response is fetched expectedly.
could you do `git pull` & re-run `npm run dev` ?

HisashiHasebe commented:
@yabe-diverta 
Umm, still happened. 
Only me?
https://diverta.gyazo.com/d559c8083eba95415e1d97a3afdc2e57

VortexExpansion committed:
fix cache problem

VortexExpansion commented:
@HisashiHasebe @yabe-diverta san

Was also happening for me.
Just cache problem. Fixed it, will revert it later after it stabilizes. Sorry for direct committing Abe san 🙇 

Please check now and confirm @HisashiHasebe san.

VortexExpansion committed:
fix class warnings

VortexExpansion committed:
hide console statements

HisashiHasebe commented:
@VortexExpansion 
Thank you! 
I can see the site .
https://diverta.gyazo.com/efa4cc72db9fb7a5ed6e63c7ee70c15e

@yabe-diverta 
Is it OK to set `{ cache: "no-store" }` ?
https://github.com/diverta/front_next_media/pull/5/commits/7fdb646fcdf5ababfed51ee6ed1cee77ac0d4b99

yabe-diverta commented:
@VortexExpansion @HisashiHasebe 
Hmm I'm not still fully understanding what's happening here, maybe DataCache provokes loading stale data? :thinking:
For now it's acceptable to merge but need to add an issue to resolve this.
